File: performance/go/03/main.go
package main

import (
	"fmt"
	"time"
)

const N = 10000000

func benchmarkSlice() {
	start := time.Now()
	slice := make([]int, 0, N)
	for i := 0; i < N; i++ {
		slice = append(slice, i)
	}
	fmt.Println("Slice Insert:", time.Since(start))

	start = time.Now()
	for len(slice) > 0 {
		slice = slice[:len(slice)-1]
	}
	fmt.Println("Slice Remove:", time.Since(start))
}

func benchmarkMap() {
	start := time.Now()
	m := make(map[int]int)
	for i := 0; i < N; i++ {
		m[i] = i
	}
	fmt.Println("Map Insert:", time.Since(start))

	start = time.Now()
	for i := 0; i < N; i++ {
		delete(m, i)
	}
	fmt.Println("Map Remove:", time.Since(start))
}

func benchmarkSet() {
	start := time.Now()
	set := make(map[int]struct{})
	for i := 0; i < N; i++ {
		set[i] = struct{}{}
	}
	fmt.Println("Set Insert:", time.Since(start))

	start = time.Now()
	for i := 0; i < N; i++ {
		delete(set, i)
	}
	fmt.Println("Set Remove:", time.Since(start))
}

func main() {
	fmt.Printf("Benchmarking Go Data Structures with %d elements:\n", N)
	benchmarkSlice()
	benchmarkMap()
	benchmarkSet()
}

File: performance/rust/03/main.rs
use std::collections::{HashMap, HashSet};
use std::time::Instant;

const N: usize = 10_000_000;

fn benchmark_vec() {
    let start = Instant::now();
    let mut v = Vec::new();
    for i in 0..N {
        v.push(i);
    }
    let duration = start.elapsed();
    println!("Vec Insert: {:?}", duration);

    let start = Instant::now();
    for _ in 0..N {
        v.pop();
    }
    println!("Vec Remove: {:?}", start.elapsed());
}

fn benchmark_hashmap() {
    let start = Instant::now();
    let mut map = HashMap::new();
    for i in 0..N {
        map.insert(i, i);
    }
    println!("HashMap Insert: {:?}", start.elapsed());

    let start = Instant::now();
    for i in 0..N {
        map.remove(&i);
    }
    println!("HashMap Remove: {:?}", start.elapsed());
}

fn benchmark_hashset() {
    let start = Instant::now();
    let mut set = HashSet::new();
    for i in 0..N {
        set.insert(i);
    }
    println!("HashSet Insert: {:?}", start.elapsed());

    let start = Instant::now();
    for i in 0..N {
        set.remove(&i);
    }
    println!("HashSet Remove: {:?}", start.elapsed());
}

fn main() {
    println!("Benchmarking Rust Data Structures with {} elements:", N);
    benchmark_vec();
    benchmark_hashmap();
    benchmark_hashset();
}
